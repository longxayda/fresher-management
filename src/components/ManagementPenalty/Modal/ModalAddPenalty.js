import React, { useState} from "react";
import { useForm } from "react-hook-form";
import {
    Button,
    Modal,
    Form,
    Container,
    Row,
    Col
} from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { addBonusPenalty } from "redux/fresherManageSlice/penaltySlice";
import { isProcessingSelector } from "redux/selectors/penaltySelector"
const colorBonus = "#28a745";
const colorPenalty = "#dc3545";
const maxDate = new Date();

function ModalAddPenalty({ userId, notifyAdd }) {
    const {
        register,
        reset,
        handleSubmit,
        formState: { errors },
    } = useForm();

    const dispatch = useDispatch();

    const isProcessing = useSelector(isProcessingSelector);
    const onSubmit = async (data) => {
        if (!errors.date && !errors.reason) {
            const newData = {
                ...data,
                user_id: userId,
            };
            try {
                await dispatch(addBonusPenalty(newData)).unwrap();
                notifyAdd(data.type);
                handleClose();
            }
            catch (err) {
                console.log(err);
                if(err.message === "Network Error"){
                    notifyAdd("Network Error");
                }
                else{
                    notifyAdd("");
                }
            }
        }
    };
    const [show, setShow] = useState(false);
    const handleClose = () => {
        setShow(false);
        reset();
    }
    const handleShow = () => {
        setShow(true);
        setColorType(colorBonus);
    }

    const [optionScore, setOptionScore] = useState(() => {
        const optionScore = [];
        for (let i = 1; i < 101; i++) {
            optionScore.push(i);
        }
        return optionScore;
    });
    
    const buttonAddClose = {
        display: "grid",
        justifyContent: "end",
        paddingRight: "0",
        paddingLeft: "0",
    };
    const titleAdd = {
        marginTop: "0",
    };
    const [colorType, setColorType] = useState(colorBonus);

    const handleStyle = (e) => {
        const value = e.target.value;

        if (value === "Bonus") {
            setColorType(colorBonus);
        }
        else {
            setColorType(colorPenalty);
        }
    };

    return (
        <>
            <Button
                id="button-add-testing"
                className="btn-fill btn-wd bbb"
                variant="primary"
                onClick={handleShow}
                style={{ width: "210px", marginLeft: "8px" }}
            >
                <i className="fas fa-plus"
                    style={{ marginRight: "4px" }}>
                </i>
                New Bonus/Penalty
            </Button>
            <Modal
                show={show}
                onHide={handleClose}
                centered
            >
                <Modal.Header closeButton>
                    <Modal.Title style={titleAdd} className="text-primary modal-title h4">
                        ADD NEW BONUS / PENALTY
                    </Modal.Title>
                    {isProcessing && <div className="processing" ></div>}
                </Modal.Header>
                <Modal.Body>
                    <Form onSubmit={handleSubmit(onSubmit)}>
                        <Form.Group>
                            <Row>
                                <Col lg="6" md sm xs="4" >
                                    <label>Date</label>
                                    <Form.Control
                                        style={errors.date ? { border: "1px solid #dc3545" } : {}}
                                        type="date"
                                        {...register("date", {
                                            validate: {
                                                require: (value) => value !== "",
                                                overtime: (value) => new Date(value) <= maxDate,
                                            }
                                        })}
                                    >
                                    </Form.Control>
                                    {errors.date && errors.date.type === "require" && (
                                        <p style={{ color: "red" }}>⚠ This field is required</p>
                                    )}
                                    {errors.date && errors.date.type === "overtime" && (
                                        <p style={{ color: "red" }}>⚠ You cannot choose future day</p>
                                    )}
                                </Col>
                                <Col lg="3" md sm xs="4" >
                                    <label>Type</label>

                                    <div>
                                        <select
                                            {...register("type")}
                                            style={{ height: "40px", width: "88px", border: "1px solid #E3E3E3", borderRadius: "4px", color: colorType }}
                                            onChange={handleStyle}
                                        >
                                            <option value="Bonus" style={{ color: colorBonus }}>Bonus</option>
                                            <option value="Penalty" style={{ color: colorPenalty }}>Penalty</option>
                                        </select>
                                    </div>
                                </Col>
                                <Col lg="3" md sm xs="4">
                                    <label> Score </label>
                                    <div>
                                        <select
                                            {...register("score")}
                                            style={{ height: "40px", width: "88px", border: "1px solid #E3E3E3", borderRadius: "4px", color: colorType }}
                                        >
                                            {optionScore.map(score => (
                                                <option key={score} value={score}>{score}</option>
                                            ))}
                                        </select>
                                    </div>
                                </Col>
                            </Row>
                        </Form.Group>
                        <label>Reason</label>
                        <Form.Control
                            {...register("reason", {
                                validate: (value) => value !== ""
                            })}
                            style={errors.reason ? { border: "1px solid #dc3545" } : {}}
                            placeholder="Here can be write your description for this bonus/penalty"
                            as="textarea"
                            rows="4"
                            className="input formReason">
                        </Form.Control>
                        {errors.reason && <p style={{ color: "red" }}>⚠ This field is required</p>}
                        <Form.Group>
                            <Container style={{ marginTop: "25px" }}>
                                <Row>
                                    <Col lg="4" md sm xs="4" ></Col>
                                    <Col lg="4" md sm xs="4" style={buttonAddClose}>
                                        <Button
                                            type="submit"
                                            className="btn-fill btn-wd"
                                            variant="primary">
                                            Add
                                        </Button>
                                    </Col>
                                    <Col lg="4" md sm xs="4" style={buttonAddClose}>
                                        <Button
                                            className="btn-fill btn-wd"
                                            variant="secondary"
                                            onClick={handleClose}>
                                            Close
                                        </Button>
                                    </Col>
                                </Row>
                            </Container>
                        </Form.Group>
                    </Form>
                </Modal.Body>
                <Modal.Footer>
                </Modal.Footer>
            </Modal>
        </>
    )
};
export default ModalAddPenalty;