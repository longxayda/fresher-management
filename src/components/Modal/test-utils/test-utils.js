import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import manageFRReducer from "../../../redux/fresherManageSlice/fresherManagementSlice";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { manageFR: manageFRReducer },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
