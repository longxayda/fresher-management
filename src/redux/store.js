import {configureStore} from "@reduxjs/toolkit";
import attendTableOneTraineeSlice from "./attendanceManagementSlice/attendTableOneTraineeSlice";

// import ticketClassSlice from "./attendanceManagementSlice/ticketClassSlice";
import diligentSummarySlice from "./attendanceManagementSlice/diligentSummarySlice";
import leaveRequestSlice from "./attendanceManagementSlice/leaveRequestSlice";
import checkboxFRReducer from "./fresherManageSlice/checkboxFresherSlice";
import evaluationSlice from "./auditorsSlice/evaluationSlice";
import resourceSlice from "./auditorsSlice/resourceSlice";
import auditorSlice from "./auditorsSlice/auditorSlice";
import manageFRReducer from "./fresherManageSlice/fresherManagementSlice";
import bonusPenaltyReducer from "./fresherManageSlice/penaltySlice";

import classSlice from "./classManagementSlice/classSlice";
import classDetailsSlice from "./classManagementSlice/classDetailsSlice";
import learningPathSlice from "./classManagementSlice/learningPathSlice";
import syllabusSlice from "./classManagementSlice/syllabusSlice";
import moduleSlice from "./classManagementSlice/moduleSlice";
import topicSlice from "./classManagementSlice/topicSlice";

import scoreSlice from "./fresherManageSlice/scoreSlice";
import topicMarkSlice from "./fresherManageSlice/topicMarkSlice";
import scheduleSlice from "./auditorsSlice/scheduleSlice";
import createNewAuditSlice from "./auditorsSlice/createNewAuditSlice";
import adminManageAttendanceReducer from "./attendanceManagementSlice/adminManageAttendanceSlice";
import adminTicketReducer from "./attendanceManagementSlice/adminTicketSlice";

import traineeAttendanceTableSlice from "./attendanceManagementSlice/traineeAttendanceTableSlice";
import authSlice from "./authSlice/authSlice";
import roleSlice from "./roleSlice";
import UserListSlice from "./UserListSlices/UserListSlice";
import attendanceSlice from "./attendanceManagementSlice/attendanceSlice";
import traineeUserProfileSlice from "./userProfileSlice/traineeUserProfileSlice";
import adminUserProfileSlice from "./userProfileSlice/adminUserProfileSlice";

import AddNewReportTableSlice from "components/WeeklyReportTables/addNewReportTableSlice";
import DetailReportTableSlice from "components/WeeklyReportTables/detailReportTableSlice";
import ReportListTableSlice from "components/WeeklyReportTables/reportListTableSlice";
import searchSlice from "components/Search/searchSlice";
import ReportModalSlice from "components/ReportModal/reportModalSlice";

import HistoryReportListTableSlice from "components/MonthlyReportTables/historyReportListTableSlice";
import coursesReportSlice from "../pages/MonthlyReportPage/MonthlyHistoryReport/courses/coursesReportSlice";
import financeReportSlice from "../pages/MonthlyReportPage/MonthlyHistoryReport/finance/financeReportSlice";
import detailCoursesReportSlice from "../pages/MonthlyReportPage/MonthlyHistoryReport/courses/detailCourseReportSlice";
import detailFinanceReportSlice from "../pages/MonthlyReportPage/MonthlyHistoryReport/finance/detailFinanceReportSlice";

import DetailDeliveryTableSlice from "components/MonthlyReportTables/detailDeliveryTableSlice";
import DeliveryReportListTableSlice from "components/MonthlyReportTables/deliveryReportListTableSlice";
import NewDeliveryTableSlice from "components/MonthlyReportTables/newDeliveryTableSlice";

import OutputsReportTableSlice from "components/OutputsReportTables/outputsReportTableSlice";
import fresherReportSlice from "../pages/OutputsReportPage/fresher/fresherReportSlice";
import svttReportSlice from "../pages/OutputsReportPage/svtt/svttReportSlice";

import detailOutputsReportSlice from "../pages/OutputsReportPage/detailOutputsReportSlice";
import detailFresherReportSlice from "../pages/OutputsReportPage/fresher/detailFresherReportSlice";
import detailSVTTReportSlice from "../pages/OutputsReportPage/svtt/detailSVTTReportSlice";

import thunk from "redux-thunk";
const store = configureStore({
	reducer: {
		Filters: searchSlice,
		evaluation: evaluationSlice,
		resource: resourceSlice,
		manageFR: manageFRReducer,
		checkboxFR: checkboxFRReducer,
		attendTableOneTrainee: attendTableOneTraineeSlice,
		// traineeTabTable: TraineeTabTableSlice,
		class: classSlice,
		learningPath: learningPathSlice,
		syllabus: syllabusSlice,
		classDetails: classDetailsSlice,
		module: moduleSlice,
		topicModule: topicSlice,
		auth: authSlice,
		role: roleSlice,
		// UserList
		UserList: UserListSlice,
		schedule: scheduleSlice,
		newAudit: createNewAuditSlice,
		auditor: auditorSlice,

		// attendTable: attendTableSlice,
		// ticketClass: ticketClassSlice,
		diligentSummary: diligentSummarySlice,
		leaveRequest: leaveRequestSlice,
		adminManageAttendance: adminManageAttendanceReducer,
		adminTicket: adminTicketReducer,
		attendance: attendanceSlice,
		traineeAttendanceTable: traineeAttendanceTableSlice,
		bonusPenalty: bonusPenaltyReducer,
		score: scoreSlice,
		topic: topicMarkSlice,
		traineeUserProfile: traineeUserProfileSlice,
		adminUserProfile: adminUserProfileSlice,

		ReportList: ReportListTableSlice,
		AddNewReport: AddNewReportTableSlice,
		DetailReport: DetailReportTableSlice,
		Filters: searchSlice,
		ReportModal: ReportModalSlice,

		HistoryReportList: HistoryReportListTableSlice,
		CoursesReport: coursesReportSlice,
		FinanceReport: financeReportSlice,
		DetailCoursesReport: detailCoursesReportSlice,
		DetailFinanceReport: detailFinanceReportSlice,
		DetailDeliveryReport: DetailDeliveryTableSlice,
		DeliveryReportList: DeliveryReportListTableSlice,
		NewDeliveryReport: NewDeliveryTableSlice,

		OutputsReport: OutputsReportTableSlice,
		FresherReport: fresherReportSlice,
		SVTTReport: svttReportSlice,
		DetailFresherReport: detailFresherReportSlice,
		DetailSVTTReport: detailSVTTReportSlice,
		DetailOutputsReport: detailOutputsReportSlice,
	},
	middleware: [thunk],
});

export default store;
