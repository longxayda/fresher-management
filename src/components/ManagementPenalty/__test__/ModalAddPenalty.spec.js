import React from "react";
import { shallow, mount, render, configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import store from "redux/store";
import { Provider } from "react-redux";
import { act } from "react-dom/test-utils";
import ModalAddPenalty from "components/ManagementPenalty/Modal/ModalAddPenalty";
configure({ adapter: new Adapter() });

describe("ModalAddPenalty", () => {
    let component;
    beforeEach(() => {
        component = mount(<Provider store={store}> <ModalAddPenalty /> </Provider>);
    });

    it("should render correctly", () => {
        expect(component.html()).toMatchSnapshot();
    });

    it("render modal", () => {
        expect(component.find('Modal').at(0).prop('show')).toBe(false);
        component.find('#button-add-testing').at(0).simulate('click');
        expect(component.find('Modal').at(0).prop('show')).toBe(true);
    });
});