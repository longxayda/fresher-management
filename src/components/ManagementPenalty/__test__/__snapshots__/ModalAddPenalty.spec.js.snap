// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalAddPenalty should render correctly 1`] = `" <button type=\\"button\\" id=\\"button-add-testing\\" style=\\"width: 210px; margin-left: 8px;\\" class=\\"btn-fill btn-wd bbb btn btn-primary\\"><i class=\\"fas fa-plus\\" style=\\"margin-right: 4px;\\"></i>New Bonus/Penalty</button> "`;
